File: frontend/src/types/analysis.ts
/**
 * TypeScript interfaces for analysis types
 */

// Query types for datasets
export type QueryType = 'embedded_sql' | 'stored_procedure' | 'shared_dataset'

// Visual types
export type VisualType =
  | 'tablix'
  | 'table'
  | 'matrix'
  | 'chart'
  | 'gauge'
  | 'map'
  | 'subreport'
  | 'textbox'
  | 'image'
  | 'rectangle'
  | 'line'
  | 'list'

// Expression categories
export type ExpressionCategory =
  | 'field_reference'
  | 'simple_aggregate'
  | 'complex_aggregate'
  | 'lookup'
  | 'custom_code'
  | 'running_value'
  | 'row_number'
  | 'previous'
  | 'unknown'

// Expression conversion categories
export type ExpressionConversionCategory = 'auto' | 'partial' | 'manual'

// TODO item categories
export type TodoCategory =
  | 'stored_procedure'
  | 'expression'
  | 'subreport'
  | 'custom_code'
  | 'unsupported_visual'

// TODO priorities
export type TodoPriority = 'high' | 'medium' | 'low'

// Conversion status
export type ConversionStatus = 'green' | 'yellow' | 'red'

// Report classification
export type ReportClassification = 'Tabular' | 'Analytical' | 'Mixed' | 'Complex'

// Dataset parameter
export interface DatasetParameter {
  name: string
  data_type?: string
  default_value?: string
}

// Dataset field
export interface DatasetField {
  name: string
  data_type?: string
  source_field?: string
}

// Dataset feature
export interface DatasetFeature {
  name: string
  query_type: QueryType
  stored_procedure_name?: string
  command_text?: string
  data_source_name?: string
  parameter_count: number
  field_count: number
  parameters: DatasetParameter[]
  fields: DatasetField[]
}

// Grouping info
export interface GroupingInfo {
  name: string
  expression?: string
  is_recursive: boolean
}

// Visual feature
export interface VisualFeature {
  type: VisualType
  name: string
  dataset_name?: string
  row_groups: number
  column_groups: number
  has_recursive_group: boolean
  nested_item_count: number
  row_group_details: GroupingInfo[]
  column_group_details: GroupingInfo[]
  subreport_path?: string
}

// Expression feature
export interface ExpressionFeature {
  expression: string
  category: ExpressionCategory
  location: string
  item_name?: string
  function_calls: string[]
}

// Layout feature
export interface LayoutFeature {
  page_width?: string
  page_height?: string
  page_width_inches?: number
  page_height_inches?: number
  orientation: string
  has_header: boolean
  has_footer: boolean
  header_height?: string
  footer_height?: string
  column_count: number
  left_margin?: string
  right_margin?: string
  top_margin?: string
  bottom_margin?: string
}

// Custom code function
export interface CustomCodeFunction {
  name: string
  parameters: string[]
  is_public: boolean
  line_count: number
}

// Analysis features
export interface AnalysisFeatures {
  rdl_version: string
  report_name?: string
  report_description?: string
  author?: string
  datasets: DatasetFeature[]
  visuals: VisualFeature[]
  expressions: ExpressionFeature[]
  layout?: LayoutFeature
  custom_code?: string
  custom_code_functions: CustomCodeFunction[]
  report_parameters: DatasetParameter[]
  data_sources: string[]
  dataset_count: number
  stored_procedure_count: number
  visual_count: number
  expression_count: number
  subreport_count: number
  running_value_count: number
  custom_code_function_count: number
  parameter_count: number
  chart_count: number
  table_count: number
  matrix_count: number
  map_count: number
  gauge_count: number
  has_custom_code: boolean
  has_stored_procedures: boolean
  has_subreports: boolean
  has_recursive_groups: boolean
  has_lookup_expressions: boolean
  has_running_values: boolean
}

// Penalty item
export interface PenaltyItem {
  category: string
  item_name?: string
  penalty_percent: number
  reason: string
}

// Score breakdown
export interface ScoreBreakdown {
  base_score: number
  penalties: PenaltyItem[]
  final_score: number
}

// Code analysis expression
export interface ExpressionAnalysis {
  expression: string
  category: ExpressionConversionCategory
  location: string
  item_name?: string
  reason: string
  suggested_dax?: string
  pattern_matched?: string
}

// TODO item
export interface TodoItem {
  id: number
  analysis_id: number
  title: string
  category: TodoCategory
  priority: TodoPriority
  location: string
  item_name?: string
  guidance: string
  original_content?: string
  is_resolved: boolean
  resolved_at?: string
  resolved_by?: number
  created_at?: string
}

// TODO list summary
export interface TodoListSummary {
  total_count: number
  high_priority_count: number
  medium_priority_count: number
  low_priority_count: number
  resolved_count: number
  unresolved_count: number
  completion_percentage: number
}

// TODO list response
export interface TodoListResponse {
  items: TodoItem[]
  summary: TodoListSummary
  high_priority_items: TodoItem[]
  medium_priority_items: TodoItem[]
  low_priority_items: TodoItem[]
  by_category: Record<string, TodoItem[]>
}

// Empty TODO list response
export interface EmptyTodoListResponse {
  message: string
  can_proceed_to_conversion: boolean
  summary: TodoListSummary
}

// Full analysis result (extended from backend response)
export interface AnalysisResult {
  id: number
  report_path: string
  report_name: string
  analyzed_at: string
  classification: ReportClassification | null
  score: number | null
  status: ConversionStatus | null
  features: AnalysisFeatures | null
  penalties: ScoreBreakdown | null
  todo_items: TodoItem[] | null
  analysis_duration_ms: number | null
}

// Type guard for TODO list responses
export function isTodoListResponse(
  response: TodoListResponse | EmptyTodoListResponse
): response is TodoListResponse {
  return 'items' in response && Array.isArray(response.items)
}

File: backend/app/schemas/analysis.py
"""Analysis Pydantic schemas for RDL feature extraction."""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class QueryType(str, Enum):
    """Type of query in a dataset."""

    EMBEDDED_SQL = "embedded_sql"
    STORED_PROCEDURE = "stored_procedure"
    SHARED_DATASET = "shared_dataset"


class VisualType(str, Enum):
    """Type of visual element in a report."""

    TABLIX = "tablix"
    TABLE = "table"
    MATRIX = "matrix"
    CHART = "chart"
    GAUGE = "gauge"
    MAP = "map"
    SUBREPORT = "subreport"
    TEXTBOX = "textbox"
    IMAGE = "image"
    RECTANGLE = "rectangle"
    LINE = "line"
    LIST = "list"


class ExpressionCategory(str, Enum):
    """Category of an expression based on complexity."""

    FIELD_REFERENCE = "field_reference"
    SIMPLE_AGGREGATE = "simple_aggregate"
    COMPLEX_AGGREGATE = "complex_aggregate"
    LOOKUP = "lookup"
    CUSTOM_CODE = "custom_code"
    RUNNING_VALUE = "running_value"
    ROW_NUMBER = "row_number"
    PREVIOUS = "previous"
    UNKNOWN = "unknown"


class DatasetParameter(BaseModel):
    """A parameter used in a dataset query."""

    name: str
    data_type: str | None = None
    default_value: str | None = None


class DatasetField(BaseModel):
    """A field returned by a dataset."""

    name: str
    data_type: str | None = None
    source_field: str | None = None


class DatasetFeature(BaseModel):
    """Features extracted from a report dataset."""

    name: str
    query_type: QueryType
    stored_procedure_name: str | None = None
    command_text: str | None = None
    data_source_name: str | None = None
    parameter_count: int = 0
    field_count: int = 0
    parameters: list[DatasetParameter] = Field(default_factory=list)
    fields: list[DatasetField] = Field(default_factory=list)


class GroupingInfo(BaseModel):
    """Information about grouping in a Tablix."""

    name: str
    expression: str | None = None
    is_recursive: bool = False


class VisualFeature(BaseModel):
    """Features extracted from a visual element."""

    type: VisualType
    name: str
    dataset_name: str | None = None
    row_groups: int = 0
    column_groups: int = 0
    has_recursive_group: bool = False
    nested_item_count: int = 0
    row_group_details: list[GroupingInfo] = Field(default_factory=list)
    column_group_details: list[GroupingInfo] = Field(default_factory=list)
    subreport_path: str | None = None  # For subreports


class ExpressionFeature(BaseModel):
    """Features extracted from an expression."""

    expression: str
    category: ExpressionCategory
    location: str  # XPath or description of where found
    item_name: str | None = None
    function_calls: list[str] = Field(default_factory=list)


class CustomCodeFunction(BaseModel):
    """A function defined in custom VB.NET code."""

    name: str
    parameters: list[str] = Field(default_factory=list)
    is_public: bool = True
    line_count: int = 0


class LayoutFeature(BaseModel):
    """Layout features of the report."""

    page_width: str | None = None  # e.g., "8.5in"
    page_height: str | None = None  # e.g., "11in"
    page_width_inches: float | None = None
    page_height_inches: float | None = None
    orientation: str = "Portrait"  # Portrait or Landscape
    has_header: bool = False
    has_footer: bool = False
    header_height: str | None = None
    footer_height: str | None = None
    column_count: int = 1
    left_margin: str | None = None
    right_margin: str | None = None
    top_margin: str | None = None
    bottom_margin: str | None = None


class AnalysisFeatures(BaseModel):
    """Complete feature extraction from an RDL file."""

    # Metadata
    rdl_version: str
    report_name: str | None = None
    report_description: str | None = None
    author: str | None = None

    # Extracted features
    datasets: list[DatasetFeature] = Field(default_factory=list)
    visuals: list[VisualFeature] = Field(default_factory=list)
    expressions: list[ExpressionFeature] = Field(default_factory=list)
    layout: LayoutFeature | None = None

    # Custom code
    custom_code: str | None = None
    custom_code_functions: list[CustomCodeFunction] = Field(default_factory=list)

    # Report parameters (not dataset parameters)
    report_parameters: list[DatasetParameter] = Field(default_factory=list)

    # Data sources
    data_sources: list[str] = Field(default_factory=list)

    # Summary counts for quick reference
    dataset_count: int = 0
    stored_procedure_count: int = 0
    visual_count: int = 0
    expression_count: int = 0
    subreport_count: int = 0
    running_value_count: int = 0
    custom_code_function_count: int = 0
    parameter_count: int = 0
    chart_count: int = 0
    table_count: int = 0
    matrix_count: int = 0
    map_count: int = 0
    gauge_count: int = 0

    # Feature flags
    has_custom_code: bool = False
    has_stored_procedures: bool = False
    has_subreports: bool = False
    has_recursive_groups: bool = False
    has_lookup_expressions: bool = False
    has_running_values: bool = False

    def to_legacy_dict(self) -> dict[str, Any]:
        """Convert to legacy dict format for backward compatibility with scoring.

        This maintains compatibility with the existing _calculate_score and
        _classify_report functions during the transition period.
        """
        return {
            "data_sources": len(self.data_sources),
            "datasets": self.dataset_count,
            "parameters": self.parameter_count,
            "tables": self.table_count,
            "matrices": self.matrix_count,
            "charts": self.chart_count,
            "subreports": self.subreport_count,
            "custom_code": self.has_custom_code,
            "expressions": self.expression_count,
            "stored_procedures": self.stored_procedure_count,
            "has_grouping": any(
                v.row_groups > 0 or v.column_groups > 0 for v in self.visuals
            ),
            "has_sorting": False,  # TODO: detect sorting
            "has_filters": False,  # TODO: detect filters
            "page_breaks": 0,  # TODO: detect page breaks
            "images": sum(1 for v in self.visuals if v.type == VisualType.IMAGE),
            "rectangles": sum(1 for v in self.visuals if v.type == VisualType.RECTANGLE),
            "textboxes": sum(1 for v in self.visuals if v.type == VisualType.TEXTBOX),
            "lines": sum(1 for v in self.visuals if v.type == VisualType.LINE),
        }


class RDLParseError(Exception):
    """Exception raised when RDL parsing fails."""

    def __init__(
        self,
        message: str,
        line_number: int | None = None,
        column: int | None = None,
        details: str | None = None,
    ):
        self.message = message
        self.line_number = line_number
        self.column = column
        self.details = details
        super().__init__(self.message)

    def __str__(self) -> str:
        parts = [self.message]
        if self.line_number is not None:
            parts.append(f"line {self.line_number}")
        if self.column is not None:
            parts.append(f"column {self.column}")
        if self.details:
            parts.append(f"({self.details})")
        return " - ".join(parts)
